igvc_perception/src/projection/main.cpp

```diff
@@ -0,0 +1,276 @@
+ * @param mapped A list of pixels that have already been mapped to x,y,z points
+ * @param cloud_list A list of pointers to point clouds, one for each lidar channel
+ */
+void image_interpolation(std::vector<std::vector<pcl::PointXYZ>> &xyz_img, std::vector<cv::Point2d> mapped,
+                         std::vector<pcl::PointCloud<pcl::PointXY>::Ptr> cloud_list)
+{
+  // KDtree for each lidar channel
+  std::vector<pcl::KdTreeFLANN<pcl::PointXY>> kdtree_list(16);
+  for (int i = 0; i < 16; i++)
+  {
+    if (!cloud_list[i]->empty())
+    {
+      kdtree_list[i].setInputCloud(cloud_list[i]);
+    }
+  }
+// Interpolate z and compute x and y for each white pixel of image
+#pragma omp parallel
+#pragma omp for collapse(2)
+  for (unsigned int r = 0; r < xyz_img.size(); r++)
+  {
+    for (unsigned int c = 0; c < xyz_img[0].size(); c++)
+    {
+      if (cv_img.at<uint8_t>(r, c) == 255 &&
+          !(std::find(mapped.begin(), mapped.end(), cv::Point2d(c, r)) != mapped.end()))
+      {
+        pcl::PointXY point = { (float)c, (float)r };
+        pcl::PointXY nearest1 = point;
+        pcl::PointXY nearest2 = point;
+        find_nearest_neighbors(point, kdtree_list, nearest1, nearest2);
+
+        cv::Point2d cv_point = cv::Point2d(c, r);
+        double inter_z = interpolate_z(nearest1, nearest2, cv_point, xyz_img);
+        xyz_img[point.y][point.x].z = inter_z;
+
+        // Find ray, and intersection of ray at z = inter_z to determine x and y.
+
+        cv::Point3d cv_proj_ray = cam_model.projectPixelTo3dRay(cv_point);
+        tf::Vector3 proj_ray = tf::Vector3(cv_proj_ray.x, cv_proj_ray.y, cv_proj_ray.z);
+        tf::Matrix3x3 rotate_matrix = transform_cam_to_base.getBasis();
+        tf::Vector3 translation = transform_cam_to_base.getOrigin();
+        tf::Vector3 trans_ray = rotate_matrix * proj_ray;
+        double scale = (inter_z - translation.getZ()) / (trans_ray.getZ());
+        xyz_img[point.y][point.x].x = trans_ray.getX() * scale + translation.getX();
+        xyz_img[point.y][point.x].y = trans_ray.getY() * scale + translation.getY();
+      }
+    }
+  }
+}
+
+void image_callback(const sensor_msgs::ImageConstPtr &msg)
+{
+  cv::Mat original_img = cv_bridge::toCvCopy(msg, "mono8")->image;
+  cv::resize(original_img, cv_img, cv::Size(resize_width, resize_height), 0, 0);
+}
+
+void lidar_pointcloud_callback(const pcl::PointCloud<velodyne_pointcloud::PointXYZIR>::ConstPtr &msg)
+{
+  std::vector<cv::Point2d> mapped;
+  std::vector<pcl::PointCloud<pcl::PointXY>::Ptr> kdtree_cloud_list(16);
+  for (int i = 0; i < 16; i++)
+  {
+    kdtree_cloud_list[i] = std::unique_ptr<pcl::PointCloud<pcl::PointXY>>(new pcl::PointCloud<pcl::PointXY>);
+  }
+  std::vector<std::vector<pcl::PointXYZ>> xyz_img(cv_img.rows, std::vector<pcl::PointXYZ>(cv_img.cols));
+  pcl::PointCloud<pcl::PointXYZ> line_cloud;
+  // Calculate which lidar points are in the fov of the camera
+  for (auto point = msg->points.begin(); point != msg->points.end(); point++)
```

Comment:
again, use const reference if you're not mutating `point`

Reply:
I have `point++` which would prevent making it a const reference. Is there a better way to iterate through the pointcloud?

Reply:
you can use range-for loops, which is equivalent to this and something you used above as long as you're not doing things like finding indices by subtracting iterators or something.

`for (const auto& point : points)`